libs/ktem/ktem/reasoning/rewoo.py

```diff
@@ -0,0 +1,422 @@
+import html
+import logging
+from difflib import SequenceMatcher
+from typing import AnyStr, Generator, Optional, Type
+
+from ktem.llms.manager import llms
+from ktem.reasoning.base import BaseReasoning
+from ktem.utils.render import Render
+from langchain.text_splitter import CharacterTextSplitter
+from pydantic import BaseModel, Field
+
+from kotaemon.agents import (
+    BaseTool,
+    GoogleSearchTool,
+    LLMTool,
+    RewooAgent,
+    WikipediaTool,
+)
+from kotaemon.base import BaseComponent, Document, HumanMessage, Node, SystemMessage
+from kotaemon.llms import ChatLLM, PromptTemplate
+
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_PLANNER_PROMPT = (
+    "You are an AI agent who makes step-by-step plans to solve a problem under the "
+    "help of external tools. For each step, make one plan followed by one tool-call, "
+    "which will be executed later to retrieve evidence for that step.\n"
+    "You should store each evidence into a distinct variable #E1, #E2, #E3 ... that "
+    "can be referred to in later tool-call inputs.\n\n"
+    "##Available Tools##\n"
+    "{tool_description}\n\n"
+    "##Output Format (Replace '<...>')##\n"
+    "#Plan1: <describe your plan here>\n"
+    "#E1: <toolname>[<input here>] (eg. Search[What is Python])\n"
+    "#Plan2: <describe next plan>\n"
+    "#E2: <toolname>[<input here, you can use #E1 to represent its expected output>]\n"
+    "And so on...\n\n"
+    "##Your Task##\n"
+    "{task}\n\n"
+    "##Now Begin##\n"
+)
+
+DEFAULT_SOLVER_PROMPT = (
+    "You are an AI agent who solves a problem with my assistance. I will provide "
+    "step-by-step plans(#Plan) and evidences(#E) that could be helpful.\n"
+    "Your task is to briefly summarize each step, then make a short final conclusion "
+    "for your task. Give answer in {lang}.\n\n"
+    "##My Plans and Evidences##\n"
+    "{plan_evidence}\n\n"
+    "##Example Output##\n"
+    "First, I <did something> , and I think <...>; Second, I <...>, "
+    "and I think <...>; ....\n"
+    "So, <your conclusion>.\n\n"
+    "##Your Task##\n"
+    "{task}\n\n"
+    "##Now Begin##\n"
+)
+
+
+class DocSearchArgs(BaseModel):
+    query: str = Field(..., description="a search query as input to the doc search")
+
+
+class DocSearchTool(BaseTool):
```

Comment:
Duplicated implementation of DocSearchTool

Reply:
https://github.com/Cinnamon/kotaemon/issues/60